src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+
+                    // If there is only one type in the unmanagedData, and that type is one of the 4 special names known to the
+                    // compiler, then we treat the CallKind of the method as if it were that calling convention. Otherwise, we
+                    // treat the CallKind as Unmanaged and do comparison from the calling convention types expected in the function
+                    // pointer.
+                    if (unmanagedCallersOnlyData.CallingConventionTypes.Count == 1)
+                    {
+                        switch (unmanagedCallersOnlyData.CallingConventionTypes.Single().Name)
+                        {
+                            case "CallConvCdecl":
+                                _ = checkConventionAgainstActual(Cci.CallingConvention.CDecl, in expectedConvention);
+                                continue;
+                            case "CallConvStdcall":
+                                _ = checkConventionAgainstActual(Cci.CallingConvention.Standard, in expectedConvention);
+                                continue;
+                            case "CallConvThiscall":
+                                _ = checkConventionAgainstActual(Cci.CallingConvention.ThisCall, in expectedConvention);
+                                continue;
+                            case "CallConvFastcall":
+                                _ = checkConventionAgainstActual(Cci.CallingConvention.FastCall, in expectedConvention);
+                                continue;
+                        }
+                    }
+
+                    if (!checkConventionAgainstActual(Cci.CallingConvention.Unmanaged, in expectedConvention))
                     {
-                        results[i] = new MemberResolutionResult<TMember>(
-                            result.Member, result.LeastOverriddenMember,
-                            MemberAnalysisResult.WrongCallingConvention());
+                        return;
+                    }
+
+                    // Now, we make sure that the actual expected types match up
+                    switch (expectedConvention.CallingConventionTypes, unmanagedCallersOnlyData.CallingConventionTypes)
+                    {
+                        case ({ IsEmpty: true }, { IsEmpty: true }):
+                            return;
+
+                        case ({ IsEmpty: true }, _) or (_, { IsEmpty: true }):
+                        case ({ Count: var expectedCount }, { Count: var actualCount }) when expectedCount != actualCount:
+                            results[i] = makeWrongCallingConvention(result);
+                            return;
+
+                    }
+
+                    foreach (var expectedModifier in expectedConvention.CallingConventionTypes)
+                    {
+                        if (!unmanagedCallersOnlyData.CallingConventionTypes.Contains(((CSharpCustomModifier)expectedModifier).ModifierSymbol))
```

Comment:
Consider using:
```
if (expectedConvention.CallingConventionTypes.Any(
    (expectedModifier, types) => !types.Contains((CSharpCustomModifier)expectedModifier).ModiferSymbol),
    unmanagedCallersOnlyData.CallingConventionTypes))
{
    ...
}
```

Reply:
We don't have this extension method on ImmutableHashSet today, but I can add one.

---
In reply to: [488848549](https://github.com/dotnet/roslyn/pull/47651#discussion_r488848549) [](ancestors = 488848549)

Reply:
When I suggested using `Any()`, I assumed `CallingConventionTypes` was an `ImmutableArray<>`. Sorry for the confusion. If we don't already have such an extension method, it's probably not worth adding one.

---
In reply to: [490551337](https://github.com/dotnet/roslyn/pull/47651#discussion_r490551337) [](ancestors = 490551337,488848549)

Reply:
Yeah, it has to be a hash set because we don't consider duplicates of the same attribute.

---
In reply to: [490558994](https://github.com/dotnet/roslyn/pull/47651#discussion_r490558994) [](ancestors = 490558994,490551337,488848549)